File: src/main/java/leetcode/p075/Solution.java
package leetcode.p075;

class Solution {

    public static void main(String[] args){
        int [] a = {1,2,0};
        Solution sol = new Solution();
        sol.sortColors(a);
        for (int x: a)
            System.out.println(x);
    }

    public void sortColors(int[] nums) {

        int zero = 0;
        int two = nums.length - 1;

        for (int i = 0; i < nums.length && i <= two; ){
            int ele = nums[i];

            if (ele == 0){
                nums[i] = nums[zero];
                nums[zero] = ele;
                zero = zero + 1;
                i++;
            }
            else {
                if (ele == 1){
                    i++;
                } else {
                    //ele == 2
                    nums[i] = nums[two];
                    nums[two] = ele;
                    two = two - 1;
                }
            }
        }
    }
}

File: scala/src/leetcode/p075/Solution.scala
package leetcode.p075

object Solution {
  def sortColors(nums: Array[Int]): Unit = {

    var i = 0
    var z = 0 // 0th ele
    var t = nums.size // last ele

    while (i < t) {
      nums(i) match {
        case 0 => {
          if (i > z) {
            swap(i, z, nums)
            z = z + 1
          } else {
            i = i + 1
            z = z + 1
          }
        }
        case 1 => {
          i = i + 1
        }
        case 2 => {
          swap(i, t - 1, nums)
          t = t - 1
        }
      }
    }
  }
  
  def swap(a: Int, b: Int, arr: Array[Int]) = {
    val tmp = arr(a)
    arr(a) = arr(b)
    arr(b) = tmp
  }
}
